src/storage/mod.rs

```diff
@@ -1847,47 +1847,34 @@ impl<E: Engine, L: LockManager, F: KvFormat> Storage<E, L, F> {
         }
     }
 
-    fn get_causal_ts(ts_provider: &Option<Arc<CausalTsProviderImpl>>) -> Result<Option<TimeStamp>> {
-        if let Some(p) = ts_provider {
-            match p.get_ts() {
-                Ok(ts) => Ok(Some(ts)),
-                Err(e) => Err(box_err!("Fail to get ts: {}", e)),
-            }
-        } else {
-            Ok(None)
-        }
-    }
+    async fn check_causal_ts_flushed(ctx: &mut Context) -> Result<()> {
+        if F::TAG == ApiVersion::V2 {
+            let snap_ctx = SnapContext {
+                pb_ctx: ctx,
+                ..Default::default()
+            };
+            let snapshot = Self::with_tls_engine(|engine| Self::snapshot(engine, snap_ctx)).await?;
 
-    async fn get_raw_key_guard(
-        ts_provider: &Option<Arc<CausalTsProviderImpl>>,
-        concurrency_manager: ConcurrencyManager,
-    ) -> Result<Option<KeyHandleGuard>> {
-        // NOTE: the ts cannot be reused as timestamp of data key.
-        // There is a little chance that CDC will acquired a timestamp for resolved-ts
-        // just between the Self::get_causal_ts & concurrency_manager.lock_key,
-        // which violate the constraint that resolve-ts should not be larger
-        // than timestamp of captured data.
-        let ts = Self::get_causal_ts(ts_provider)?;
-        if let Some(ts) = ts {
-            let raw_key = vec![api_version::api_v2::RAW_KEY_PREFIX];
-            // Make keys for locking by RAW_KEY_PREFIX & ts. RAW_KEY_PREFIX to avoid
-            // conflict with TiDB & TxnKV keys, and ts to avoid collision with
-            // other raw write requests. Ts in lock value to used by CDC which
-            // get maximum resolved-ts from concurrency_manager.global_min_lock_ts
-            let encode_key = ApiV2::encode_raw_key(&raw_key, Some(ts));
-            let key_guard = concurrency_manager.lock_key(&encode_key).await;
-            let lock = Lock::new(LockType::Put, raw_key, ts, 0, None, 0.into(), 1, ts);
-            key_guard.with_lock(|l| *l = Some(lock));
-            Ok(Some(key_guard))
-        } else {
-            Ok(None)
+            if !snapshot.ext().is_max_ts_synced() {
+                return Err(Error::from(txn::Error::from(
+                    TxnError::MaxTimestampNotSynced {
+                        region_id: ctx.get_region_id(),
+                        start_ts: TimeStamp::zero(),
+                    },
+                )));
+            }
+            let term = snapshot.ext().get_term();
+            if let Some(term) = term {
+                ctx.set_term(term.get());
```

Comment:
Why we need `set_term` here ? And there is no this block in atomic commands.

Reply:
For ABA, if event `leader tranfer from A to B` & `leader tranfer from B to A` happend between after `check_causal_ts_flushed` and before `raft propose`. For atomic commands, ref[ here](https://github.com/tikv/tikv/blob/master/src/storage/txn/scheduler.rs#L599)

Reply:
Got it.